File: countArrangement.java
// Time Complexity : O(n!)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
class Main {
    // to maintain a count of beautiful arrangement
    private static int count;

    public static int countArrangement(int n) {
        count = 0;
        boolean[] visited = new boolean[n + 1];
        backTrack(1, n, visited);
        return count;
    }

    private static void backTrack(int index, int n, boolean[] visited) {

        // base case
        if (index > n) {
            count++;
            return;
        }

        // main logic
        for (int i = 1; i <= n; i++) {
            // if element is not visited and it is beautiful then we make it visited and
            // recurse
            if (!visited[i] && (index % i == 0 || i % index == 0)) {
                // action
                visited[i] = true;
                // recurse
                backTrack(index + 1, n, visited);
                // backtrack
                visited[i] = false;

            }
        }
    }

    public static void main(String[] args) {
        System.out.println(countArrangement(3));
    }
}

File: countArrangement.py
# Time Complexity : Exponential
# Space Complexity : O(N)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Your code here along with comments explaining your approach

class Solution:
    def countArrangement(self, n: int) -> int:
        self.count = 0
        visited = set()
        index = 1
        #1 -> N
        self.backtrack(n, index,  visited)
        
        return self.count
    
    def backtrack(self, n, index, visited):
        if index > n:
            self.count += 1
        
        #logic
        for i in range(1, n+1):
            if i not in visited and (i % index == 0 or index % i == 0):
                #action
                visited.add(i)
                #recurse
                self.backtrack(n, index+1, visited)
                #backtrack
                visited.remove(i)
    